Blog-projects/Widget-window-in-Qt-Quick-app/fontcontrolswidgetsform.cpp

```diff
@@ -0,0 +1,48 @@
+#include "fontcontrolswidgetsform.h"
+#include "ui_fontcontrolswidgetsform.h"
+
+FontControlsWidgetsForm::FontControlsWidgetsForm(QWidget *parent)
+    : QWidget(parent)
+    , ui(new Ui::FontControlsWidgetsForm)
+{
+    ui->setupUi(this);
+    connect(ui->lineEdit, &QLineEdit::textEdited, this, &FontControlsWidgetsForm::lineEditTextChanged);
+    connect(ui->fontComboBox, &QFontComboBox::currentFontChanged, this, &FontControlsWidgetsForm::currentFontChanged);
+    connect(ui->pushButton, &QPushButton::clicked, this, &FontControlsWidgetsForm::buttonClicked);
+}
+
+FontControlsWidgetsForm::~FontControlsWidgetsForm()
+{
+    delete ui;
```

Comment:
The problem with saying "This code follows best practices" in the commit log, is that I'll keep spotting anything that doesn't actually follow best practices ;)

The best practice here is to turn `ui` into a std::unique_ptr, and turn this destructor into
`FontControlsWidgetsForm::~FontControlsWidgetsForm() = default;`
(stil here in the .cpp file - don't move it to the header because there the Ui class isn't fully defined and some compilers won't like it)

Reply:
I'm getting a good refresher in the process. Changed to a unique pointer.

As for the destructor, I noticed that if I use a unique pointer and don't set the default destructor or don't set it the way you indicated, with GCC I will get something like `error: invalid application of ‘sizeof’ to incomplete type ‘FontControlsWidgetsForm’` is that what you were referring to?

Reply:
Yes, exactly. If the destructor isn't there or is defaulted in the header file, when gcc generates it (in the context of the header file), it doesn't know the size of the type so it doesn't know how much memory to deallocate.